File: python/binary_search/boj2470.py
import sys

input = sys.stdin.readline

# 산성: 1~ 10억 양의 정수 알칼리성: -1 ~ -10억 : int범위
# 합해서 0에 가까운 용액
# [-2,4,-99,-1,98] => -99 + 98 = -1 => 0에 가장 가깝다
# 숫자를 하나씩 돌면서 합이 0에 가까운 수를 찾는다 => 가까운수? = 정렬
# a랑 합이 0이다 = -a => -a에 가장 가까운 정수를 찾기


# ----------------
def lower_bound(a, l, r, x):
    res = r + 1
    while l <= r:
        mid = (l + r) // 2
        if a[mid] >= x:
            res = mid
            r = mid - 1
        else:
            l = mid + 1
    return res


n = int(input())
a = sorted(list(map(int, input().split())))
best_sum = 1 << 31
v1, v2 = 0, 0
for l in range(n - 1):
    cand = lower_bound(a, l + 1, n - 1, -a[l])
    if l < cand - 1 and abs(a[l] + a[cand - 1]) < best_sum:
        best_sum = abs(a[l] + a[cand - 1])
        v1, v2 = a[l], a[cand - 1]
    if cand < n and abs(a[l] + a[cand]) < best_sum:
        best_sum = abs(a[l] + a[cand])
        v1, v2 = a[l], a[cand]

print(v1, v2)

File: javaalgointerview/rhs/binarysearch/BOJ2470.java
package rhs.binarysearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

/*
두개를 합쳐 0에 가장 가깝게 만들자
 */
public class BOJ2470 {

	static FastReader scan = new FastReader();
	static StringBuilder sb = new StringBuilder();

	static int N;
	static int[] A;

	static void input() {
		N = scan.nextInt();
		A = new int[N+1];
		for (int i = 1; i <= N; i++) {
			A[i] = scan.nextInt();
		}
	}

	static int lowerBound(int[] A, int L, int R, int X) {
		int res = R + 1;
		while (L <= R) {
			int mid = (L+R) / 2;
			if (A[mid] >= X) {
				res = mid;
				R = mid - 1;
			} else {
				L = mid + 1;
			}
		}
		return res;
	}

	static void pro() {
		Arrays.sort(A, 1, N+1);

		int bestSum = Integer.MAX_VALUE;
		int v1 = 0, v2 = 0;
		for (int left = 1; left <= N; left++) {
			int cand = lowerBound(A, left + 1, N, -A[left]);

			if (left < cand - 1 && Math.abs(A[left] + A[cand - 1]) < bestSum) {
				bestSum = Math.abs(A[left] + A[cand - 1]);
				v1 = A[left];
				v2 = A[cand - 1];
			}

			if (cand <= N && Math.abs(A[left] + A[cand]) < bestSum) {
				bestSum = Math.abs(A[left] + A[cand]);
				v1 = A[left];
				v2 = A[cand];
			}
		}
	}

	public static void main(String[] args) {

	}

	static class FastReader {
		BufferedReader br;
		StringTokenizer st;

		public FastReader() {
			br = new BufferedReader(new InputStreamReader(System.in));
		}

		public String next() {
			while (st == null || !st.hasMoreElements()) {
				try {
					st = new StringTokenizer(br.readLine());
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
			return st.nextToken();
		}

		int nextInt() {
			return Integer.parseInt(next());
		}

		long nextLong() {
			return Long.parseLong(next());
		}
	}
}
